rclcpp/include/rclcpp/node.hpp

```diff
@@ -936,17 +938,69 @@ class Node : public std::enable_shared_from_this<Node>
    * of the latter things,
    * rclcpp::exceptions::ParameterModifiedInCallbackException will be thrown.
    *
-   * There may only be one callback set at a time, so the previously set
-   * callback is returned when this method is used, or nullptr will be returned
-   * if no callback was previously set.
+   * The callback functions must remain valid as long as the
+   * returned unique pointer is valid.
+   * The returned unique pointer can be promoted to a shared version.
+   *
+   * Releasing the returned unique pointer unregisters the callback.
+   * `remove_on_set_parameters_callback` can also be used.
+   *
+   * All the registered callbacks are called when a parameter is set.
+   * The order of the callbacks is not important.
```

Comment:
Is it not? What is the order called? What happens if one of them fails (returns a result with success set to false), do the others that have not been called still get called? If so, how do you combine the results if they disagree or more than one fails?

Reply:
I've updated the comment, I hope it's clearer now.

When rewriting it, I realized about the following:
Users are using callbacks for updating configurations when they are returning a successful result (e.g.: https://github.com/clalancette/velodyne/blob/34af494b6a8caad6e97378026e353ca9cfdaa97f/velodyne_pointcloud/src/conversions/convert.cc#L186-L219). In case other callback was registered, it may be executed later and reject the result, and the prior callback would have updated the configuration.

I wrote this thinking that each callback will check a disjoint group of parameters, but I don't know if that assumption is reasonable or not. If it is reasonable, I would just document how it works. If not, maybe having pre and post callbacks is a good idea.

Reply:
Based on #769, I think it's a reasonable assumption.
But, there's maybe an use case where is not.

Reply:
> I wrote this thinking that each callback will check a disjoint group of parameters, but I don't know if that assumption is reasonable or not.

I don't think that it is. The assumption of the callback must be that the values may still not be set due to another callback rejecting them. That's the only safe thing to do.

There is another set of callbacks "on change" which get called after parameters are added, updated, or removed. Currently it uses a topic and a subscription to generate these callbacks, but should be changed to use some other local mechanism (like directly calling a callback from within set parameter methods): https://github.com/ros2/rclcpp/issues/609

By the way, this pull request should probably somehow mention this issue https://github.com/ros2/rclcpp/issues/722 which it is addressing.